### PR TITLE
OPEN-29 Add insert method

### PR DESCRIPTION
**Pull request by mawagusura-psqr**
The original `save` method uses `INSERT ... ON DUPLICATE KEY UPDATE ...` by default when persisting entities (for MySQL at least).
This causes problems when having unique constraints on fields different than the PK : MySQL will update the preexisting row except the PK and the unique key, leading to incoherent data.
To counter this, I added an `insert` method that will perform a regular insert, and throw a `DuplicateKeyException` exception if a unique constraint is not matched.

This new method uncovered some issues in the TestDoubleResolver that Anton has fixed in this same PR (and also in ran).

### CODE REVIEW

File: valqueries-automapper/src/main/java/com/valqueries/automapper/ValqueriesAccessDataLayerImpl.java [hunk 1]
```diff
@@ -175,25 +176,62 @@ public CrudUpdateResult save(T t) {
...
+			return () -> 0;
+		}
+		CompoundColumnizer<O> columnizer = new CompoundColumnizer<O>(genericFactory, mappingHelper, ts, sqlNameFormatter);
+		String sql = dialect.getInsert(columnizer, oClass);
+		UpdateResult result;
+		try {
+			result = tx.update(sql, columnizer); 
+		} catch (Exception e){
+			// TODo should we restrict the exceptions that we flag as duplicate key ?
+			throw new ValqueriesDuplicateKeyException(e);
```
Comment (mikkelep) [inline valqueries-automapper/src/main/java/com/valqueries/automapper/ValqueriesAccessDataLayerImpl.java:190]: How do we know that this is a duplicate key problem, and not e.g. a transaction conflict?

Comment (mawagusura-psqr) [reply valqueries-automapper/src/main/java/com/valqueries/automapper/ValqueriesAccessDataLayerImpl.java:190]: We don't, hence the TODO comment. I will change the exception to be more generic, the reason shouldn't matter too much anyway.

**Review by mikkelep — APPROVED** [review]

File: valqueries-automapper/src/main/java/com/valqueries/automapper/ValqueriesAccessDataLayerImpl.java (hunk 1 above)
Comment (mawagusura-psqr) [reply valqueries-automapper/src/main/java/com/valqueries/automapper/ValqueriesAccessDataLayerImpl.java]: After talking to Anton, I tried to see if we could catch some specific jdbc exceptions only. However, the SQL Server exceptions do not inherit the jdbc ones, making it impossible to catch in the repository layer. To avoid any big refactoring, I just generalized the exception, as it makes sense to tell the outer layer that the insert fail regardless of the reason.

File: valqueries-test/src/test/java/com/valqueries/automapper/AutoMapperBaseTests.java [hunk 2]
```diff
@@ -1223,6 +1225,265 @@ public void update_withoutPreviousRecord() {
...
+	public void insert_happy() {
+		Car model = factory.get(Car.class);
+		model.setId(UUID.randomUUID());
+		model.setTitle("Muh");
+		model.setBrand(Brand.Porsche);
+		model.setCreatedAt(ZonedDateTime.now().withZoneSameInstant(ZoneOffset.UTC).truncatedTo(ChronoUnit.SECONDS));
+		carRepository.doRetryableInTransaction(tx -> carRepository.insert(tx, model));
+
+		Optional<Car> actualOptional = carRepository.get(model.getId());
+		Car actual = actualOptional.orElseThrow(RuntimeException::new);
```
Comment (teofilomonteiro) [inline valqueries-test/src/test/java/com/valqueries/automapper/AutoMapperBaseTests.java:1239]: this is a minor thing but why you throwing RuntimeException? why is this necessary? by default if you try to get it will throw an `NoSuchElementException`. If its to better assert this, why not something like:

```suggestion
		if(!actualOptional.isPresent()){
			fail("Car id not found in database");
		}
		Car actual = actualOptional.get();
```

**Review by teofilomonteiro — APPROVED** [review]
Apart from the comment that I left seems fine
